### PR TITLE
Extension/associations

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: app/models/order.rb [hunk 1]
```diff
@@ -49,4 +50,8 @@ def charge!(pay_type_params)
       raise payment_result.error
     end
   end
+
+  def total_amount
+    line_items.sum { |item| item.total_price }
```
Comment (himanshumishra31) [inline app/models/order.rb]: line_items.sum(&:total_price) 

Use shorthand notation

File: app/models/line_item.rb [hunk 2]
```diff
@@ -1,10 +1,12 @@
 class LineItem < ApplicationRecord
   belongs_to :order, optional: true
   belongs_to :product
-  belongs_to :cart, optional: true
-
-  validates :product_id, uniqueness: { scope: :cart_id }
-
+  belongs_to :cart, optional: true, counter_cache: true
+  
+  validates :product_id, uniqueness: { scope: :cart_id }, if: :cart
```
Comment (himanshumishra31) [inline app/models/line_item.rb:6]: there should be a validation also for product_id with scope order_id

File: app/models/product.rb [hunk 3]
```diff
@@ -7,9 +7,10 @@ def validate_each(record, attribute, value)
 class Product < ApplicationRecord
   after_initialize :set_title_default
   before_save :set_discount_price_default
-  has_many :line_items
+  has_many :line_items, dependent: :restrict_with_error
   has_many :orders, through: :line_items
-  before_destroy :ensure_not_referenced_by_any_line_item
+  has_many :carts, through: :line_items
+  # before_destroy :ensure_not_referenced_by_any_line_item
```
Comment (himanshumishra31) [inline app/models/product.rb:13]: why commented?

File: app/models/line_item.rb [hunk 4]
```diff
@@ -1,10 +1,12 @@
 class LineItem < ApplicationRecord
   belongs_to :order, optional: true
   belongs_to :product
-  belongs_to :cart, optional: true
-
-  validates :product_id, uniqueness: { scope: :cart_id }
-
+  belongs_to :cart, optional: true, counter_cache: true
```
Comment (himanshumishra31) [inline app/models/line_item.rb:4]: Also study how counter_cache finds column in other model

Also how can i change column name for counter cache

File: app/models/cart.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
 class Cart < ApplicationRecord
   has_many :line_items, dependent: :destroy
+  has_many :products, through: :line_items
+  has_many :enabled_products, -> {where enabled: true}, through: :line_items, source: :product
```
Comment (himanshumishra31) [inline app/models/cart.rb]: create scope for `where enabled: true ` in products.
Also use parentheses around where for better readability

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -68,6 +68,16 @@ def destroy
     redirect_to users_url, notice: exception.message
   end
 
+  def orders
+    @user = User.find(session[:user_id])
+    @orders = @user.orders
+  end
+
+  def line_items
+    @user = User.find(session[:user_id])
```
Comment (himanshumishra31) [inline app/controllers/users_controller.rb]: Always handle failed cases also. If session[:user_id] is not found, your endpoint will give 500.
Endpoint should return proper message with proper status code.

**Comment by himanshumishra31** [comment]
Comments added.
Also try out all options from api.

File: app/models/product.rb (hunk 3 above)
Comment (rishabh-vinsol) [reply app/models/product.rb:13]: Line number 10 has the same implementation and this was the change that was asked in the question

File: app/models/line_item.rb (hunk 4 above)
Comment (rishabh-vinsol) [reply app/models/line_item.rb:4]: Counter_cache finds the column like <model_name>_count in the table that it belongs to. 
If we want to specify the count column we can use the definition like `counter_cache: :name_of_column`

File: app/models/user.rb [hunk 7]
```diff
@@ -20,9 +22,9 @@ class Error < StandardError
   private
 
   def check_admin_email
-    if email == ADMIN_EMAIL
+    if User.find(id).email == ADMIN_EMAIL
```
Comment (himanshumishra31) [inline app/models/user.rb]: use dirty objects

File: db/migrate/20230403093313_add_user_ref_to_orders.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class AddUserRefToOrders < ActiveRecord::Migration[7.0]
+  def change
+    add_reference :orders, :user, foreign_key: true
```
Comment (himanshumishra31) [inline db/migrate/20230403093313_add_user_ref_to_orders.rb:3]: `foreign_key` not required here

File: config/routes.rb [hunk 9]
```diff
@@ -7,7 +7,10 @@
   end
   get "sessions/create"
   get "sessions/destroy"
-  resources :users
+  resources :users do
+    get "orders"
```
Comment (himanshumishra31) [inline config/routes.rb]: generic convention...use symbol
get :orders

File: db/migrate/20230403093313_add_user_ref_to_orders.rb (hunk 8 above)
Comment (rishabh-vinsol) [reply db/migrate/20230403093313_add_user_ref_to_orders.rb:3]: Sir, by default foreign_key defaults to false

**Pull request merged**